### PR TITLE
add args to wc_create_new_customer to autogenerate password/username

### PR DESCRIPTION
**Pull request by haszari**
Related: https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2377 https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2851#issuecomment-664092646

When an account is created, a username and password can be automatically generated. This is a really helpful feature, as it makes signing up that much easier. This is especially relevant when shoppers are signing up as part of a checkout flow. If we can keep this simple it's easier for shoppers and better for merchants :)

In WooCommerce, there are two store options in `WooCommerce > Settings > Accounts & Privacy` that control this behaviour:

- `When creating an account, automatically generate an account username for the customer based on their name, surname or email`
- `When creating an account, automatically generate an account password`

Screen Shot 2020-07-27 at 3 20 56 PM

In the new checkout block, we're keen to always generate the username and password when creating accounts, to reduce the number of fields/text boxes between a shopper and completing checkout. (At least, for the initial version of the block.)

Currently there's no way to programmatically create a user account with an autogenerated username and password - the core routine `wc_create_new_customer()` always refers to the options. This PR is a proposal that the routine should _also_ allow autogenerate behaviour based on arguments to the function. 

Note that the option values may also be consulted elsewhere in Woo core code. From my research it looks like the password option is consulted in the template for the new account email; I don't think this is necessary, as the template has an explicit parameter indicating a generated password.

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### How to test the changes in this Pull Request:
The goal of this PR is that existing behaviour is unaffected, and the API routine is more flexible for use by blocks. So the testing steps here are regression testing.

1. Test creating an account via various flows (signup form, create account via checkout, invite user from admin, etc).
2. Repeat tests with different combinations of autogenerate options.
3. Ensure that new accounts are created correctly with appropriate username and password (user or automatically provided), and that the new account email content is relevant, useful and correct in all cases.

To test the more flexible interface to `wc_create_new_customer()`, you could try calling the routine in `wp shell` or a custom plugin, or via an open PR in blocks plugin (TBC!).

e.g. 

```php
wc_create_new_customer(
  'shopper@shopaholic.org',
  '',
  '', 
  [
    'autogenerate_password' => true,
    'autogenerate_username' => true,
  ]
);
```
### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

Haven't considered tests yet, but if we do proceed with this I think it's a good opportunity to tighten up this area with some test coverage and isolate some of the code responsible for creating accounts. cc @ObliviousHarmony  🧪 

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### CODE REVIEW

File: includes/wc-user-functions.php [hunk 1]
```diff
@@ -897,7 +900,7 @@ function wc_update_user_last_active( $user_id ) {
 	if ( ! $user_id ) {
 		return;
 	}
-	update_user_meta( $user_id, 'wc_last_active', (string) strtotime( date( 'Y-m-d', time() ) ) );
+	update_user_meta( $user_id, 'wc_last_active', (string) strtotime( gmdate( 'Y-m-d', time() ) ) );
```
Comment (haszari) [inline includes/wc-user-functions.php:903]: This change was prompted by phpcs linter.

**Review by ObliviousHarmony — CHANGES_REQUESTED** [review]
Hello @haszari,

My first thought when reviewing this change was that we should pass a generated `$username` and `$password` to this function if we want to override the option's behavior. The caveat there though is that the `woocommerce_created_customer` action will have an incorrect value for `$password_generated` based on the assumption it wasn't. I think with that in mind this is a reasonable change. 

As far as testing goes, we've actually got really good coverage of this function in the `tests/legacy/unit-tests/customer/functions.php` file! Note that there is now a failing test in Travis due to this change. You could add this case to the existing test function and that should suffice.

If I was to do anything here myself I'd likely refactor the function into a few different classes, but I won't impose my insanity on others 😄

**Comment by haszari** [comment]
Thanks for the review @ObliviousHarmony !

I've marked this PR as blocked for now. We're looking at implementing this differently in blocks – the plan is to add a new create account routine (PHP api) that ignores the options and always creates a new account as follows:

- Automatically generates a username.
- Emails the user to set the initial password.

Of course we may leverage existing Woo core code to implement some of this new flow. 

See the linked blocks [PR](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2377) and [issue](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2851#issuecomment-664092646) for more details (work in progress). pca54o-75-p2

> As far as testing goes, we've actually got really good coverage of this function in the tests/legacy/unit-tests/customer/functions.php file! Note that there is now a failing test in Travis due to this change. You could add this case to the existing test function and that should suffice.

Aha, thanks for highlighting that. As an exercise for myself, I've pushed a change which hopefully fixes the failure. Side note: I don't know how to run Woo core unit tests locally – I use docker based environments for dev, not sure what needs setting up for the unit tests. (I see there are [docs](https://github.com/woocommerce/woocommerce/tree/master/tests) but I think they are focused more on VVV use case.)

> If I was to do anything here myself I'd likely refactor the function into a few different classes, but I won't impose my insanity on others 😄

You actually raise an interesting point. When I add the new create account flow, I'm implementing essentially a new woo core PHP api, in blocks plugin initially. So it would be good if we build the new flow using similar patterns/approach as you're using for new code in core, to ease the path for merging into core in future. So I might ping you for a review on the blocks side :)

**Comment by haszari** [comment]
Closing this – we're handling this by using a token-based set password flow. https://github.com/woocommerce/woocommerce/issues/27754

**Pull request closed without merging**